libfwupdplugin/fu-cfi-device.c

```diff
@@ -34,6 +40,10 @@ enum { PROP_0, PROP_FLASH_ID, PROP_LAST };
 
 #define GET_PRIVATE(o) (fu_cfi_device_get_instance_private(o))
 
+#define FU_CFI_DEVICE_PAGE_SIZE_DEFAULT	  0x100
+#define FU_CFI_DEVICE_SECTOR_SIZE_DEFAULT 0x1000
+#define FU_CFI_DEVICE_BLOCK_SIZE_DEFAULT  0x10000
```

Comment:
There values look relatively sane to me as far as I know.

I have currently the chips below
- [GD25Q32C](https://www.gigadevice.com/datasheet/gd25q32c/)
- [GD25Q40C](https://www.gigadevice.com/datasheet/gd25q40c/)
- [FM25F01](http://eng.fmsh.com/nvm/FM25F01_ds_eng.pdf)
- [FM25W04](http://eng.fmsh.com/nvm/FM25W04_ds_eng.pdf)
- [MX25L3236](https://www.macronix.com/Lists/Datasheet/Attachments/7416/MX25L3236F,%203V,%2032Mb,%20v1.2.pdf)
- [MX25V2033F](https://www.macronix.com/Lists/Datasheet/Attachments/7443/MX25V2033F,%202.5V,%202Mb,%20v1.0.pdf)
- [GD25Q20C](https://www.gigadevice.com/datasheet/gd25q20c/)
- [GD25Q32C](https://www.gigadevice.com/datasheet/gd25q32c/)

There are using page-size 0x100 (Page Program 02h), sector-size 0x1000 (Sector Erase 20h) and 32k-block-size 0x8000 (Block Erase 32K 52h) and 64k-block-size 0x10000 (Block Erase 64K D8h).

So both 0x8000 and 0x10000 values are good candidate for block-size. I guess we should pick up the value that match the default command number for block erase; if 52h -> 0x8000; if D8h -> 0x10000.

Reply:
> I have currently the chips below

You couldn't add all those to cfi.quirk please? I think it's better to be 100% explicit.

Reply:
Hum, there is not default value set for block erase yet (`CfiDeviceCmdBlockErase`).

I see only one plugin that references block erase and it uses a 64k erase block command `0xd8` (`SIO_SPI_CMD_64K_BLOCK_ERASE`).

Reply:
What does the jdec says about the `BE` command? what is the default value? or maybe there is already two `BE` commands.

Also, some documentations mention about half block for 32k bocks. therefore, I think `0x10000` is my favorite for now.

Reply:
> > I have currently the chips below
> 
> You couldn't add all those to cfi.quirk please? I think it's better to be 100% explicit.

Do you mean adding this kind of change?

``` diff
# GigaDevice
[CFI\FLASHID_C8]
Vendor = GigaDevice
[CFI\FLASHID_C840]
Name = GD25Qxxx
CfiDeviceCmdChipErase = 0xC7
CfiDeviceCmdSectorErase = 0x20
+CfiDeviceCmdBlockErase = 0xD8
+CfiDeviceBlockSize = 0x10000
+CfiDeviceSectorSize = 0x1000
+CfiDevicePageSize = 0x100
+
+[CFI\FLASHID_C84012]
+Name = GD25Q20C
+FirmwareSizeMax = 0x40000
+
+[CFI\FLASHID_C84016]
+Name = GD25Q32C
+FirmwareSizeMax = 0x400000
```

I have not because I was not sure actually.

Reply:
Also, I would like (and I am not sure) that `[CFI\FLASHID_C84016]` will use the unset properties from `[CFI\FLASHID_C840]`. and IIRC, it was not; at least when I tested.

Reply:
> Do you mean adding this kind of change?

Yup, exactly like that.

> will use the unset properties

FLASHID_C84016 overrides FLASHID_C840 which overrides FLASHID_C8 -- i.e. most specific to least.

Reply:
perfect. so I will make this changes and test.

Reply:
Added some to the my existing PR #4473; I can make a separate PR for GD and MXIC chips.

Reply:
Yes please, thanks!

Reply:
#4482